pkg/client/leaderelection/leaderelection.go

```diff
@@ -0,0 +1,217 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package leaderelection implements leader election of a set of endpoints.
+// It uses an annotation in the endpoints object to store the record of the
+// election state.
+package leaderelection
```

Comment:
If you can decouple the mechanics and leverage through an interface I'm ok with with calling it leaderelection b/c I could back end against my own algorithm which may/may-not talk to some other facility.  This way the plumbing is the same, and implementation specifics are hidden.  In this case you would hide the implementation of endpoints annotation such that controllers and scheduler only pass in a "key".   This 1 layer of indirection would make it general enough such that someone else could write an algo and all the plumbing would be the same. 

How does this sound?  Happy to do 1:1 hangout if what I said was confusing.